numpy/_core/src/multiarray/arraywrap.c

```diff
@@ -0,0 +1,308 @@
+/*
+ * Definitions for dealing with array-wrap or array-prepare.
+ */
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+#define _UMATHMODULE
+
+#include <Python.h>
+
+#include "numpy/arrayobject.h"
+#include "numpy/npy_3kcompat.h"
+#include "get_attr_string.h"
+
+#include "arraywrap.h"
+#include "multiarraymodule.h"
+
+
+/*
+ * Find the array wrap or array prepare method that applies to the inputs.
+ * outputs should NOT be passed, as they are considered individually while
+ * applying the wrapping.
+ *
+ * @param nin number of inputs
+ * @param inputs Original input objects
+ * @param out_wrap Set to the python callable or None (on success).
+ * @param out_wrap_type Set to the type belonging to the wrapper.
+ */
+NPY_NO_EXPORT int
+npy_find_array_wrap(
+        int nin, PyObject *const *inputs,
+        PyObject **out_wrap, PyObject **out_wrap_type)
+{
+    PyObject *wrap = NULL;
+    PyObject *wrap_type = NULL;
+
+    double priority = 0;  /* silence uninitialized warning */
+
+    for (int i = 0; i < nin; i++) {
+        PyObject *obj = inputs[i];
+        if (PyArray_CheckExact(obj)) {
+            if (wrap == NULL || priority < 0) {
+                Py_INCREF(Py_None);
+                Py_XSETREF(wrap, Py_None);
+                priority = 0;
+            }
+        }
+        else if (PyArray_IsAnyScalar(obj)) {
+            if (wrap == NULL || priority < NPY_SCALAR_PRIORITY) {
+                Py_INCREF(Py_None);
+                Py_XSETREF(wrap, Py_None);
+                priority = NPY_SCALAR_PRIORITY;
+            }
+        }
+        else {
+            PyObject *new_wrap = PyArray_LookupSpecial_OnInstance(obj, npy_ma_str_array_wrap);
```

Comment:
I guess this can be inside the `if` statement below, since `new_wrap` is not used if the priority is too low. Though that means two lookups if `__array_wrap__` does not exist, so maybe what you have is better. If you keep it, though, I think you need to `Py_DECREF new_wrap` in an `else` clause (as it is a new reference).

Reply:
Ouch, thanks!  I agree with moving looking nicer, but then decided that maybe `__array_priority__` without `__array_wrap__` is actually not that uncommon, so just added the decref.